### PR TITLE
feat: [IDP-1814]: IDP catalog entities as Harness entities

### PR DESCRIPTION
**Pull request by sathish-soundarapandian**
## Describe your changes
IDP catalog entities as Harness entities

## Checklist
- [x] I've documented the changes in the PR description.
- [x] I've tested this change either in PR or local environment.
- [x] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

### Rebase Instructions:
To rebase your branch onto the latest changes in the target branch, simply use the following command in a comment: `/rebase`

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/56679)

### CODE REVIEW

**Review by vigneswara-propelo — COMMENTED** [review]
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/56679)* status: 95 file(s) to review. 0 of 2 approvals obtained. 

___
*[`idp-service/src/main/java/io/harness/idp/backstage/service/impl/BackstageServiceImpl.java` line 59 at r1](https://reviewable.io/reviews/harness/harness-core/56679#-NlWX47B-nsYnXjzDH6p:-NlWX47B-nsYnXjzDH6q:b-v8dqug) ([raw file](https://github.com/harness/harness-core/blob/ee51f17abb8c957f4c890681fb65f30b7930de64/idp-service/src/main/java/io/harness/idp/backstage/service/impl/BackstageServiceImpl.java#L59)):*
> ```Java
>     try {
>       log.info("Syncing IDP catalog entities as Harness entities for accountIdentifier = {}", accountIdentifier);
>       String url = String.format("%s/idp/api/catalog/entities", accountIdentifier);
> ```

We are trying to fetch all the entities from backstage without any filters. This would take several minutes if an account has huge number of entities. 
___
*[`idp-service/src/main/java/io/harness/idp/backstage/service/impl/BackstageServiceImpl.java` line 137 at r1](https://reviewable.io/reviews/harness/harness-core/56679#-NlW_bGg2PerjFDHEdtB:-NlW_bGg2PerjFDHEdtC:b-mchm1z) ([raw file](https://github.com/harness/harness-core/blob/ee51f17abb8c957f4c890681fb65f30b7930de64/idp-service/src/main/java/io/harness/idp/backstage/service/impl/BackstageServiceImpl.java#L137)):*
> ```Java
>       backstageCatalogEntity.setAccountIdentifier(accountIdentifier);
>       backstageCatalogEntity.setEntityUid(entityUid);
>       backstageCatalogEntity.setYaml("");
> ```

do we need this?



_Code quote:_
```Java
backstageCatalogEntity.setYaml("");
```
___
*[`idp-service/src/main/java/io/harness/idp/backstage/service/impl/BackstageServiceImpl.java` line 163 at r1](https://reviewable.io/reviews/harness/harness-core/56679#-NlWYmaF-6yh_8LqtS9g:-NlWYmaF-6yh_8LqtS9h:b83ydqj) ([raw file](https://github.com/harness/harness-core/blob/ee51f17abb8c957f4c890681fb65f30b7930de64/idp-service/src/main/java/io/harness/idp/backstage/service/impl/BackstageServiceImpl.java#L163)):*
> ```Java
> 
>   private void handleCreateOrUpdateAction(String accountIdentifier, String entityUid, String action) {
>     Object response = getGeneralResponse(backstageResourceClient.getCatalogEntityByName(accountIdentifier, entityUid));
> ```

Are we constructing this entityUid as "kind"/"namespace"/"entity-name" ? I could see this construction for DB fetch and not for catalog API call



_Code quote:_
```Java
entityUid
```

**Review by vigneswara-propelo — COMMENTED** [review]
> Reviewed 2 of 95 files at r1.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/56679)* status: 95 file(s) to review. 0 of 2 approvals obtained. 

___
*[`idp-service/src/main/java/io/harness/idp/backstage/service/impl/BackstageServiceImpl.java` line 72 at r1](https://reviewable.io/reviews/harness/harness-core/56679#-NlWyvuLEb_42W9FzXX8:-NlWyvuLEb_42W9FzXX9:by6gla2) ([raw file](https://github.com/harness/harness-core/blob/ee51f17abb8c957f4c890681fb65f30b7930de64/idp-service/src/main/java/io/harness/idp/backstage/service/impl/BackstageServiceImpl.java#L72)):*
> ```Java
>   public void sync(String accountIdentifier, String entityUid, String action, String syncMode) {
>     switch (syncMode) {
>       case "sync":
> ```

nit: we could make this constant as its being used in other place also



_Code quote:_
```Java
sync
```

**Review by vikyathharekal — COMMENTED** [review]
> Reviewed 22 of 95 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/56679)* status: 93 file(s) to review. 0 of 2 approvals obtained. 

___
*[`953-events-api/src/main/proto/io/harness/eventsframework/schemas/idp/backstage_catalog.proto` line 13 at r1](https://reviewable.io/reviews/harness/harness-core/56679#-NlWcMBzEKA6JIgLbtvd:-NlWcMBzEKA6JIgLbtve:b1qyxes) ([raw file](https://github.com/harness/harness-core/blob/ee51f17abb8c957f4c890681fb65f30b7930de64/953-events-api/src/main/proto/io/harness/eventsframework/schemas/idp/backstage_catalog.proto#L13)):*
> ```proto
> option go_package = "953_events_api";
> 
> message BackstageCatalogEntityChangeDto {
> ```

What happens to others listening to the entity_crud stream? They might be converting this proto to EntityChangeDTO
___
*[`idp-service/contracts/openapi/v1/openapi.yaml` line 1410 at r1](https://reviewable.io/reviews/harness/harness-core/56679#-NlWdRklAR-Ak9S9i-sj:-NlWdRklAR-Ak9S9i-sk:bc3i67b) ([raw file](https://github.com/harness/harness-core/blob/ee51f17abb8c957f4c890681fb65f30b7930de64/idp-service/contracts/openapi/v1/openapi.yaml#L1410)):*
> ```YAML
>       security:
>         - x-api-key: []
>   /v1/backstage-harness/sync/for-account/{accountIdentifier}:
> ```

Do we need this `for-account`? 
`/v1/backstage-harness/sync/{accountIdentifier}` should be enough right?



_Code quote:_
```YAML
for-account
```

**Review by vikyathharekal — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/56679)* status: 93 file(s) to review. 0 of 2 approvals obtained. 

___
*[`idp-service/contracts/openapi/v1/openapi.yaml` line 1410 at r1](https://reviewable.io/reviews/harness/harness-core/56679#-NlWdRklAR-Ak9S9i-sj:-NlX-LDz4ZgDNLSfCmPn:b-ineyq3) ([raw file](https://github.com/harness/harness-core/blob/ee51f17abb8c957f4c890681fb65f30b7930de64/idp-service/contracts/openapi/v1/openapi.yaml#L1410)):*
<details><summary><i>Previously, vikyathharekal (Vikyath Harekal) wrote…</i></summary><blockquote>

Do we need this `for-account`? 
`/v1/backstage-harness/sync/{accountIdentifier}` should be enough right?
</blockquote></details>

Also account is already in the header. Can we do something with this itself?

**Review by vikyathharekal — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/56679)* status: 93 file(s) to review. 1. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`idp-service/src/main/java/io/harness/idp/backstage/service/impl/BackstageServiceImpl.java` line 72 at r1](https://reviewable.io/reviews/harness/harness-core/56679#-NlWyvuLEb_42W9FzXX8:-NlX-lMDB5c-eQRMKUzk:b-1ncq8e) ([raw file](https://github.com/harness/harness-core/blob/ee51f17abb8c957f4c890681fb65f30b7930de64/idp-service/src/main/java/io/harness/idp/backstage/service/impl/BackstageServiceImpl.java#L72)):*
<details><summary><i>Previously, vigneswara-propelo (Vigneswara) wrote…</i></summary><blockquote>

nit: we could make this constant as its being used in other place also
</blockquote></details>

Can we use the Enum itself instead of string?
___
*[`idp-service/src/main/java/io/harness/idp/backstage/service/impl/BackstageServiceImpl.java` line 145 at r1](https://reviewable.io/reviews/harness/harness-core/56679#-NlX1YfD5OQT7E1ZnFDl:-NlX1YfD5OQT7E1ZnFDm:b-ww3tr4) ([raw file](https://github.com/harness/harness-core/blob/ee51f17abb8c957f4c890681fb65f30b7930de64/idp-service/src/main/java/io/harness/idp/backstage/service/impl/BackstageServiceImpl.java#L145)):*
> ```Java
>     String namespace = isEmpty(backstageCatalogEntity.getMetadata().getNamespace())
>         ? "default"
>         : backstageCatalogEntity.getMetadata().getNamespace();
> ```

Can we use entity.metadata.uid? Backstage itself provides unique identifier right?



_Code quote:_
```Java
backstageCatalogEntity.getMetadata(
```

**Review by vikyathharekal — COMMENTED** [review]
> Reviewed 73 of 95 files at r1.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/56679)* status: 93 file(s) to review. 1. unresolved discussions. 0 of 2 approvals obtained.

**Review by sathish-soundarapandian — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/56679)* status: 93 file(s) to review. 3. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`idp-service/src/main/java/io/harness/idp/backstage/service/impl/BackstageServiceImpl.java` line 163 at r1](https://reviewable.io/reviews/harness/harness-core/56679#-NlWYmaF-6yh_8LqtS9g:-NlYT1F39aEbZXCmqAw8:b5w0ges) ([raw file](https://github.com/harness/harness-core/blob/ee51f17abb8c957f4c890681fb65f30b7930de64/idp-service/src/main/java/io/harness/idp/backstage/service/impl/BackstageServiceImpl.java#L163)):*
<details><summary><i>Previously, vigneswara-propelo (Vigneswara) wrote…</i></summary><blockquote>

Are we constructing this entityUid as "kind"/"namespace"/"entity-name" ? I could see this construction for DB fetch and not for catalog API call
</blockquote></details>

Yes. entityUid will be "kind/namespace/name". Both for DB save / fetch and catalog API call, we use the same notation.
entityUid is the name I have given for uniquely identifying an IDP entity at Harness side. idp-app when calling the API will send entityUid as "kind/namespace/name"



_Code snippet:_
```Java
  @GET(CATALOG_API + "/entities/by-name/{name}")
  Call<Object> getCatalogEntityByName(
      @Path("accountIdentifier") String accountIdentifier, @Path(value = "name", encoded = true) String name);
```

**Review by sathish-soundarapandian — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/56679)* status: 93 file(s) to review. 2. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`idp-service/src/main/java/io/harness/idp/backstage/service/impl/BackstageServiceImpl.java` line 72 at r1](https://reviewable.io/reviews/harness/harness-core/56679#-NlWyvuLEb_42W9FzXX8:-NlYYjrf0mDKHzOY8tZm:bb6un1j) ([raw file](https://github.com/harness/harness-core/blob/ee51f17abb8c957f4c890681fb65f30b7930de64/idp-service/src/main/java/io/harness/idp/backstage/service/impl/BackstageServiceImpl.java#L72)):*
<details><summary><i>Previously, vikyathharekal (Vikyath Harekal) wrote…</i></summary><blockquote>

Can we use the Enum itself instead of string?
</blockquote></details>

Done

**Review by sathish-soundarapandian — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/56679)* status: 93 file(s) to review. 3. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`idp-service/src/main/java/io/harness/idp/backstage/service/impl/BackstageServiceImpl.java` line 59 at r1](https://reviewable.io/reviews/harness/harness-core/56679#-NlWX47B-nsYnXjzDH6p:-NlYe0P1AYrr0NiAHcwk:bq8wgls) ([raw file](https://github.com/harness/harness-core/blob/ee51f17abb8c957f4c890681fb65f30b7930de64/idp-service/src/main/java/io/harness/idp/backstage/service/impl/BackstageServiceImpl.java#L59)):*
<details><summary><i>Previously, vigneswara-propelo (Vigneswara) wrote…</i></summary><blockquote>

We are trying to fetch all the entities from backstage without any filters. This would take several minutes if an account has huge number of entities.
</blockquote></details>

Yeah. I understand the concern here.
But this sync-all flow will be only one time for all accounts as part of the background migration. And if needed, it will act as force sync debug API which can be executed by harness support user.
So, thinking whether we need to put effort in implementing paginated call. We cannot use any filter because we need to fetch all the entities for sync.
Pagination we can implement, but again save to DB will be in half hanging.
By the way, does getCatalogEntities api in backstage support pagination of page and limit option ?

**Review by sathish-soundarapandian — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/56679)* status: 93 file(s) to review. 5. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`idp-service/contracts/openapi/v1/openapi.yaml` line 1410 at r1](https://reviewable.io/reviews/harness/harness-core/56679#-NlWdRklAR-Ak9S9i-sj:-NlaIOq76sxz4kNLs96m:b-vc8zq5) ([raw file](https://github.com/harness/harness-core/blob/ee51f17abb8c957f4c890681fb65f30b7930de64/idp-service/contracts/openapi/v1/openapi.yaml#L1410)):*
<details><summary><i>Previously, vikyathharekal (Vikyath Harekal) wrote…</i></summary><blockquote>

Also account is already in the header. Can we do something with this itself?
</blockquote></details>

Done. Simplified it.
But we need accountIdentifier in path param for the 2nd API, else it will conflict with the 1st one

**Review by vigneswara-propelo — COMMENTED** [review]
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/56679)* status: 93 file(s) to review. 4. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`idp-service/src/main/java/io/harness/idp/backstage/service/impl/BackstageServiceImpl.java` line 59 at r1](https://reviewable.io/reviews/harness/harness-core/56679#-NlWX47B-nsYnXjzDH6p:-NlaSnqf-K3RxxZQ2lRv:bgozbbx) ([raw file](https://github.com/harness/harness-core/blob/ee51f17abb8c957f4c890681fb65f30b7930de64/idp-service/src/main/java/io/harness/idp/backstage/service/impl/BackstageServiceImpl.java#L59)):*
<details><summary><i>Previously, sathish-soundarapandian (Sathish) wrote…</i></summary><blockquote>

Yeah. I understand the concern here.
But this sync-all flow will be only one time for all accounts as part of the background migration. And if needed, it will act as force sync debug API which can be executed by harness support user.
So, thinking whether we need to put effort in implementing paginated call. We cannot use any filter because we need to fetch all the entities for sync.
Pagination we can implement, but again save to DB will be in half hanging.
By the way, does getCatalogEntities api in backstage support pagination of page and limit option ?
</blockquote></details>

yeah that will be a little extra work from our end but a different  API called 'queryEntities' support pagination

**Review by sathish-soundarapandian — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/56679)* status: 93 file(s) to review. 5. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`953-events-api/src/main/proto/io/harness/eventsframework/schemas/idp/backstage_catalog.proto` line 13 at r1](https://reviewable.io/reviews/harness/harness-core/56679#-NlWcMBzEKA6JIgLbtvd:-NlanQ7h8Va056gir70C:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/ee51f17abb8c957f4c890681fb65f30b7930de64/953-events-api/src/main/proto/io/harness/eventsframework/schemas/idp/backstage_catalog.proto#L13)):*
<details><summary><i>Previously, vikyathharekal (Vikyath Harekal) wrote…</i></summary><blockquote>

What happens to others listening to the entity_crud stream? They might be converting this proto to EntityChangeDTO
</blockquote></details>

Done.

**Review by vikyathharekal — DISMISSED** [review]
:lgtm:

> Reviewed 2 of 2 files at r3, 1 of 2 files at r4, 2 of 2 files at r5, 3 of 3 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/56679#-:-Nlb3MHZAs-F3qfUaP1h:bnfp4nl)* status: 93 file(s) to review. 3. unresolved discussions. 1 of 2 approvals obtained.

**Review by vigneswara-propelo — APPROVED** [review]

**Review by vikyathharekal — DISMISSED** [review]
> Reviewed 14 of 14 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/56679)* status: 93 file(s) to review. 2. unresolved discussions. 1 of 2 approvals obtained.

**Review by vigneswara-propelo — DISMISSED** [review]

**Comment by sathish-soundarapandian** [comment]
trigger review, uts

**Comment by sathish-soundarapandian** [comment]
trigger ut2, ut4, ut7

**Comment by sathish-soundarapandian** [comment]
trigger ut7

**Comment by sathish-soundarapandian** [comment]
trigger uts

**Comment by sathish-soundarapandian** [comment]
trigger ut0, ut5

**Comment by sathish-soundarapandian** [comment]
trigger ut0

**Comment by sathish-soundarapandian** [comment]
trigger ut0

File: idp-service/contracts/openapi/v1/openapi.yaml [hunk 1]
```diff
@@ -1411,6 +1411,53 @@ paths:
...
+        - BackstageHarness
+      responses:
+        '200':
+          $ref: '#/components/responses/BackstageHarnessSyncEntitiesResponse'
+      operationId: backstage-harness-sync-all-accounts
+      description: Syncs Backstage Entities as Harness Entities for all IDP enabled accounts
+      x-internal: true
+      security:
+        - x-api-key: []
+  /v1/backstage-harness/sync/{accountIdentifier}:
```
Comment (mankrit-singh) [inline idp-service/contracts/openapi/v1/openapi.yaml:1427]: can we remove accountIdentifier from the path? if not can we rename it to account-id or account?

Comment (sathish-soundarapandian) [reply idp-service/contracts/openapi/v1/openapi.yaml]: Updated path to /v1/backstage-harness/sync/for-account and taking accountIdentifier from header for usage

**Review by mankrit-singh — APPROVED** [review]

**Comment by sathish-soundarapandian** [comment]
trigger uts

**Review by vikyathharekal — APPROVED** [review]
> Reviewed 2 of 2 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/56679)* status: 93 file(s) to review. 3. unresolved discussions.

**Comment by sathish-soundarapandian** [comment]
trigger review

**Comment by sandeepa-kv-jois** [comment]
Unrelated UTs are failing, going ahead with FM

**Pull request merged**